### ISSUE TITLE
How to find memory leaks?

### ISSUE THREAD

**Issue by tgulacsi** [oracle/odpi#118]
I see that my Go program (using goracle, which uses ODPI-C) consumes more-and-more memory over time servicing requests (calling stored procedures).

The Go runtime sees only around 60Mb acquired from the OS, but RSS is around 800Mb.

I've running the program with DPI_DEBUG_LEVEL=32, collecting the allocations,
and here are those that are not freed (according to the debug messages) [1]:
[nonfreed.log](https://github.com/oracle/odpi/files/3807617/nonfreed.log)

Most of them are OCI allocations. Are those not freed, or only not printed here?
Shall I try with other DPI_DEBUG_LEVEL? Or just skip the OCI allocations/frees?

Some are ODPI allocations. How can I find my non-releases?
In theory all requests prepare a statement, bind variables, execute the statement,
read the variables, and release the statement. What shall I check?

The raw messages:
[odpi.log.gz](https://github.com/oracle/odpi/files/3807625/odpi.log.gz)

---
[1]:
```
#!/bin/bash
fn="${1:-odpi.log.gz}"
gzip -dc "$fn" | grep -E "$(comm -23 \
        <(gzip -dc "$fn" | sed -ne '/allocated/ s/^.* \(0x[^ ]*\).*$/\1/p' | sort -u -S 100m --compress-program=gzip --parallel=4) \
        <(gzip -dc "$fn" | sed -ne '/freed/ s/^.* 0x/0x/p' | sort -u -S 100m --compress-program=gzip --parallel=4) | \
        tr "\n" '|' | sed -s 's/|$//')"
```

**Comment by anthony-tuininga** [comment]
Can you post the original log file? You can use DPI_DEBUG_LEVEL=38 so that it includes the reference counts and the function calls or just DPI_DEBUG_LEVEL=34 for the reference counts. I have a script that processes these files written in Python which I can share if that will be helpful. It reports on which memory has not been freed.

**Comment by tgulacsi** [comment]
TL;DR user VARCHAR2 instead of CHAR.

The program is a web server, which calls stored procedures, using generated code.

Since the report, I've
  1. decreased the PL/SQL associative arrray sizes from 512 to 128 (except where needed).
  2. a) decreased the max session lifetime to [5m](https://www.unosoft.hu/tgulacsi/ws-server.odpi.to=05m.20191108.log.gz), to force the connections to be closed.
      b) decreased the max session lifetime to [60m](https://www.unosoft.hu/tgulacsi/ws-server.odpi.to=60m.20191108.log.gz), to force the connections to be closed.

Even with these changes, the production server consumes up to 900Mb, 
but the same program in the DEV environment stays below 400Mb.

The remaining difference between the DEV and PROD is that 've replaced a few "CHAR" OUT arguments with VARCHAR2, as AFAIK CHAR requires 32k, VARCHAR2 does not.

This change will go to prod in 10 days from now.


Client: 19.3.0.0.0
Server: 19.5.0.0.0 [Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production; Version 19.5.0.0.0]

The log files are from the DEV env, each about 30m timespan.
The "stress test" is by repeating what the PROD got (over 3 days) using 8 concurrent requests.

**Comment by anthony-tuininga** [comment]
So does that mean you have resolved your issue? Or did you still have questions that need to be answered? If so, can you provide some code? I understand Go well enough to be able to follow it without difficulty -- and the SQL and PL/SQL code are probably of equal interest.

**Comment by tgulacsi** [comment]
If your script finds any discrepancies in those logs (5m or 60m), then we should investigate further - but yes, I've found a half-solution-half-workaround: use VARCHAR2 instead of CHAR in OUT parameters, and close the connections regularly.

**Comment by anthony-tuininga** [comment]
I downloaded those files and attempted to gunzip them -- but got "unexpected end of file" -- which may be due to the size of the files in question. It would be better to reduce the size of these files -- preferably by creating a standalone case that you can run repeatedly that demonstrates memory increasing continuously. That is what I do when I am trying to diagnose a memory leak. I keep reducing the test case until I find the smallest one that replicates the issue. If you're happy with what you have, though, feel free to simply close the issue!

**Comment by tgulacsi** [comment]
I cannot reproduce the memory consumption in a controlled environment.

1000 calls with one OUT CHAR: 
After start, without any calls: 18572
After 1 call: 46284
After 1000 calls: 65336

1000 calls with OUT VARCHAR2:
After start, without any calls: 19044
After 1 call: 45940
After 1000 calls: 58136

My last try is some statistics by running each function 500 times:
[rss.xlsx](https://github.com/oracle/odpi/files/3827058/rss.xlsx)

I cannot see anything in it. What shall I check? What does an outlier look like?

**Comment by anthony-tuininga** [comment]
I'm not sure what you mean by the numbers you have noted above and in the attached spreadsheet. Can you clarify? I generally create a simple script that contains a loop. Inside the loop are the series of statements that I wish to test (which should be memory neutral if everything is coded correctly). I then run that loop repeatedly and simply check the output of the "top" command (or other method that can be used to determine the RSS of the process). If I see that the memory is indeed increasing, I then use ``DPI_DEBUG_LEVEL=38`` and run the script I mentioned to determine whether ODPI-C and/or OCI is the culprit -- and if so, what function is doing it. I also quickly check to see that the number of references is being held constant and something isn't grabbing a reference and simply not releasing it -- that can frequently be the source of issues! If you have some code that you would like to have reviewed, let me know.

**Comment by tgulacsi** [comment]
Those numbers are RSS Kilobytes.
I couldn't find a tendency in those numbers - nothing really related to the number of cycles.

Each function calls a different stored procedure. The code is generated, so not really the same, but quite similar.
I've cold started the server, then called then stored procedure once, measured the memory, then 50 times, measured the memory, then another 450 times, and measured again.
This is the 3 "rss" columns.
But neither function stood out - neither the one that only returns a simple string, nor the one that returns an array of PL/SQL record...

So now I'm closing this issue, thanks for your help - maybe there's no memory leak, I'll see it on the next week, when the code is deployed in production (I cannot reproduce the memory usage on my dev server).

**Comment by anthony-tuininga** [comment]
You're welcome. Feel free to ask again if you get more information!

**Comment by tgulacsi** [comment]
Sorry to reiterate it, but I still see 1000Mb RSS at the production environment, but only under 300Mb when repeating all the same requests to the same program - at the development/testing environments.

The only difference between the production and development/testing enviroments is that the development env uses a simple Oracle (Enterprise) DB, but the prod has a (one-node) RAC.

Can RAC cause such ballooning of memory? 

I'm using pooling with GORACLE as the pool's name, minSessions=2, maxSessions=32, sessionTimeout=1m, maxLifetime=1h. About 10 sessions are show in v$session, neither of them is older than 1h.

**Comment by anthony-tuininga** [comment]
Is the memory stable but just substantially larger in production vs development environments? Or is memory increasing over time? Is the database character set the same in the production and development environments? This can make a substantial difference! You can find the database character set by issuing this query:

```sql
SELECT value AS db_charset
FROM nls_database_parameters
WHERE parameter = 'NLS_CHARACTERSET';
```

One possibility of determing the source of the differences is to set the environemnt variable ``DPI_DEBUG_LEVEL`` to the value 32 and then run one iteration of your code in production and one in development and examine the output and see what the differences are. In particular, pay attention to the memory that OCI is allocating.

One other possibility which ODPI-C does not provide (but possibly could in the future if this is found to be helpful) but OCI does, is to examine the value of the environment attribute ``OCI_ATTR_HEAPALLOC``, which gives you a 32-bit value corresponding to the amount of memory allocated for that particular OCI environment. ODPI-C creates one environment for each standalone connection and one for each pool.

**Comment by tgulacsi** [comment]
The memory is increasing to 1Gb, maxes out at about 1.3Gb. So nothing urgent here, just it bothers me that why is that so much and why can't I reproduce it.

NLS_CHARACTERSET=EE8ISO8859P2 on both systems.

I'll report bck with OCI_ATTR_HEAPALLOC.

**Comment by tgulacsi** [comment]
How can I contribute the dpiConn_getHeapAlloc and dpiPool_getHeapAlloc functions?
[getHeapAlloc.diff.txt](https://github.com/oracle/odpi/files/3864017/getHeapAlloc.diff.txt)

**Comment by anthony-tuininga** [comment]
@tgulacsi, were you able to figure out where the differences were by using this function?

I'm not sure if there is much benefit to exposing that new function directly as is, but it might make sense to output the OCI heap allocation after each ODPI-C function call when ``DPI_DEBUG_LEVEL`` is set to 32. Thoughts?

**Comment by tgulacsi** [comment]
No, I couldn't run it, it has some issues (SIGSEGV in production, though works in testing).
Printing when debugging sounds good!

**Comment by anthony-tuininga** [comment]
Ok. I'll discuss adding that with @cjbj and if he agrees we'll add that capability. If you figure out the source of the memory discrepancies, please let us know!
